xayn-ai/src/ltr/list_net/mod.rs

```diff
@@ -198,12 +234,130 @@ impl ListNet {
 
         for chunk in chunks {
             scores.extend_from_slice(
-                &self.calculate_final_scores_padded(first_chunk, Some(chunk))[chunk_size..],
+                &self.calculate_final_scores_padded(first_chunk, Some(chunk))[Self::CHUNK_SIZE..],
             );
         }
 
         scores
     }
+
+    /// Computes the gradients and loss for given inputs and target prob. dist.
+    ///
+    /// # Panics
+    ///
+    /// If inputs and relevances are not for exactly [`Self::INPUT_NR_DOCUMENTS`]
+    /// documents.
+    fn gradients_for_query(&self, sample: Sample) -> (GradientSet, f32) {
+        let Sample {
+            inputs,
+            target_prob_dist,
+        } = sample;
+        assert_eq!(inputs.shape()[0], ListNet::INPUT_NR_DOCUMENTS);
+        assert_eq!(target_prob_dist.len(), ListNet::INPUT_NR_DOCUMENTS);
+        let results = self.train_forward_pass(inputs);
+        //FIXME[followup PR] if we don't track loss in XaynNet when used in the app we don't need to calculate it.
+        let loss = kl_divergence(target_prob_dist.view(), results.prob_dist_y.view());
+        let gradients = self.train_back_propagation(inputs, target_prob_dist, results);
+        (gradients, loss)
+    }
+
+    /// Run the the forward pass of back-propagation.
+    fn train_forward_pass(&self, inputs: ArrayView2<f32>) -> ForwardPassData {
+        let (scores_y, partial_forward_pass) = self.calculate_intermediate_scores(inputs, true);
+        let prob_dist_y = self.calculate_final_scores(&scores_y);
+
+        ForwardPassData {
+            partial_forward_pass: partial_forward_pass.unwrap(),
+            scores_y,
+            prob_dist_y,
+        }
+    }
+
+    /// Run back propagation base on given inputs, target prob. dist. and forward pass data.
+    fn train_back_propagation(
+        &self,
+        inputs: ArrayView2<f32>,
+        target_prob_dist: ArrayView1<f32>,
+        forward_pass: ForwardPassData,
+    ) -> GradientSet {
+        let ForwardPassData {
+            partial_forward_pass:
+                PartialForwardPassData {
+                    dense1_y,
+                    dense1_z,
+                    dense2_y,
+                    dense2_z,
+                },
+            scores_y,
+            mut prob_dist_y,
+        } = forward_pass;
+
+        let nr_documents = inputs.shape()[0];
+
+        let p_cost_and_prob_dist = {
+            prob_dist_y -= &target_prob_dist;
+            prob_dist_y
+        };
```

Comment:
Got point, it can be simplified again (I started out with what you proposed but at some point this didn't work due to some things related to ndarray trait impl, references and views.).